src/main/java/com/squareup/javapoet/CodeBlock.java

```diff
@@ -106,41 +107,69 @@ public Builder add(String format, Object... args) {
         if (format.charAt(p) != '$') {
           nextP = format.indexOf('$', p + 1);
           if (nextP == -1) nextP = format.length();
+          formatParts.add(format.substring(p, nextP));
         } else {
-          checkState(p + 1 < format.length(), "dangling $ in format string %s", format);
-          switch (format.charAt(p + 1)) {
+          checkState(p <= format.length(), "dangling $ in format string %s", format);
```

Comment:
My mistake, changed the wrong line based upon previous comments and didn't catch it. Fixed in next check in.